### PR TITLE
Update logic of resolving SdkReferenceAssemblies.

### LINKED ISSUE #2638: Reference assemblies are located in the wrong folder on Windows
### Description

After upgrading to `5.21.0`, I can no longer run my FAKE script on Windows.

```
There was a problem while setting up the environment:
-> Could not find referenced assemblies in path: 'C:\Users\nojaf\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0\ref\net6.0', please check installed SDK and runtime versions
```

### Repro steps

> dotnet fake -v build

```
C:\Users\nojaf\Projects\Datto\network-scanner>dotnet fake -v build
runOrBuild ({ Script = None
  ScriptArguments = []
  FsiArgLine = []
  Debug = false
  NoCache = false
  RestoreOnlyGroup = false
  VerboseLevel = Verbose
  IsBuild = true })
FAKE 5 - F# Make (5.21.0) (this line is written to standard error, see https://github.com/fsharp/FAKE/issues/2066)
prepareAndRunScriptRedirect(Script: C:\Users\nojaf\Projects\Datto\network-scanner\build.fsx, fsiOptions: "")
Writing 'C:\Users\nojaf\Projects\Datto\network-scanner\.fake\build.fsx\intellisense.fsx'
Restoring with paket...
The last restore is still up to date. Nothing left to do.
Retrieving the assemblies (rid: 'win10-x64')...
Using .Net 6 assemblies
Trying to resolve runtime version from network..
Calculating the runtime graph...
Loaded runtime json from: microsoft.netcore.platforms-5.0.3
resolved runtime version: 6.0.0
Resolved referenced SDK path: C:\Users\nojaf\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0\ref\net6.0
Performance:
 - Cli parsing: 132 milliseconds
 - Packages: 1 second
   - Creating Runtime Graph: 40 milliseconds
   - Retrieve Assembly List: 1 second
 - Runtime: 2 seconds
There was a problem while setting up the environment, see standard error for details.
There was a problem while setting up the environment:
-> Could not find referenced assemblies in path: 'C:\Users\nojaf\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0\ref\net6.0', please check installed SDK and runtime versions
   StackTrace:
        at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1439.Invoke(String message) in D:\workspace\_work\1\s\src\fsharp\FSharp.Core\printf.fs:line 1439
        at Fake.Runtime.FakeRuntime.retrieveInfosUncached@114(String cacheDir, Lazy`1 paketDependenciesFile, VerboseLevel logLevel, GroupName groupName, SdkAssemblyResolver sdkAssemblyResolver, FrameworkIdentifier framework, Rid rid, Rid ridNotVersionSpecific, Lazy`1 lockFile, Lazy`1 cache, Unit unitVar0) in D:\
a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 202
        at Fake.Runtime.FakeRuntime.getKnownDependencies@312-1.Invoke(Unit unitVar0) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 312
        at Fake.Runtime.CoreCache.getCached[a](FSharpFunc`2 getUncached, FSharpFunc`2 readFromCache, FSharpFunc`2 writeToCache, FSharpFunc`2 checkCacheUpToDate) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\CoreCache.fs:line 41
        at Fake.Runtime.FakeRuntime.getKnownDependencies@310(String cacheDir, Lazy`1 paketDependenciesFile, VerboseLevel logLevel, GroupName groupName, FileInfo lockFilePath, String dependencyCacheHashFile, String dependencyCacheFile, SdkAssemblyResolver sdkAssemblyResolver, FrameworkIdentifier framework, Rid ri
d, Rid ridNotVersionSpecific, Lazy`1 lockFile, Lazy`1 cache, Lazy`1 writeIntellisenseTask, FSharpFunc`2 readFromCache, FSharpFunc`2 writeToCache, Unit unitVar0) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 311
        at Fake.Runtime.FakeRuntime.knownDependencies@320.Invoke(Unit unitVar) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 320
        at System.Lazy`1.ViaFactory(LazyThreadSafetyMode mode)
        at System.Lazy`1.ExecutionAndPublication(LazyHelper executionAndPublication, Boolean useDefaultConstructor)
        at System.Lazy`1.CreateValue()
        at System.Lazy`1.get_Value()
        at Fake.Runtime.FakeRuntime.paketCachingProvider(FakeConfig config, String cacheDir, Dependencies paketApi, Lazy`1 paketDependenciesFile, FSharpOption`1 group) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 323
        at Fake.Runtime.FakeRuntime.runScript(PrepareInfo preparedScript) in D:\a\FAKE\FAKE\src\app\Fake.Runtime\FakeRuntime.fs:line 579
        at Program.runOrBuild(RunArguments args) in D:\a\FAKE\FAKE\src\app\Fake.netcore\Program.fs:line 156

```

### Expected behavior

FAKE script runs.

### Actual behavior

The reference assemblies are being located from `C:\Users\<user>\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref`, on my machine they are stored on `C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0`.

### Known workarounds

Copying the `6.0.0` folder from `Program Files` to  `AppData\Local` solves the problem.

### Related information 

* Operating system: Windows 11
* Branch: default branch I think
* .NET Runtime, CoreCLR or Mono Version: .NET 6
* Indications of severity: FAKE is unusable.
* Version of FAKE (4.X, 5.X): 5.21

### PR DESCRIPTION
**Pull request by nojaf**
### Description

- fixes #2638

### CODE REVIEW

File: src/app/Fake.Runtime/SdkAssemblyResolver.fs [hunk 1]
```diff
@@ -124,41 +126,105 @@ type SdkAssemblyResolver(logLevel:Trace.VerboseLevel) =
...
-            )
-            |> Seq.toList
-        | false ->
-            failwithf "Could not find referenced assemblies in path: '%s', please check installed SDK and runtime versions" referenceAssembliesPath
+            dotnetRoot
+            |> Option.map (fun dotnetRoot ->
+                dotnetRoot.Directory.FullName
+                </> "packs"
+                </> "Microsoft.NETCore.App.Ref"
+                </> this.ResolveSdkRuntimeVersion()
```
Comment (nojaf) [inline src/app/Fake.Runtime/SdkAssemblyResolver.fs:211]: Not something I want to tackle in this PR but `this.ResolveSdkRuntimeVersion()` is just going to throw if no `global.json` is present. This feels wrong, could the version not be extracted from running `dotnet --version` instead if no global.json is present?

**Comment by nojaf** [comment]
Oh, and I tested this on my local project and it worked on my machine 😊.

**Comment by nojaf** [comment]
Failing tests feel unrelated,

```
Failed:  3
	Fake.DotNet.Cli.IntegrationTests.Template tests/can install and run the template/can build a buildtask-dsl inline-dependencies template
	Fake.DotNet.Cli.IntegrationTests.Template tests/can install and run the template/can build a buildtask-dsl file-dependencies template
	Fake.DotNet.Cli.IntegrationTests.Template tests/can install and run the template/fails to build a target that doesn't exist
```

**Comment by yazeedobaid** [comment]
@nojaf Thanks for the PR, the template integration tests started to fail not sure why!
This happens after we upgrade to net 6.0.101 in FAKE. I'll try to look at them later today.

Thanks

**Comment by yazeedobaid** [comment]
Tests should be fixed in [this PR](https://github.com/fsprojects/FAKE/pull/2640).

**Comment by nojaf** [comment]
Thanks @yazeedobaid, I'll rebase when https://github.com/fsprojects/FAKE/pull/2640 got merged in.
Assuming my PR goes green then, could we have a hotfix release for `fake-cli` once this is merged?

**Comment by yazeedobaid** [comment]
@nojaf Yeah sure, if all goes well then we can release this as a hotfix.

**Comment by yazeedobaid** [comment]
Should we also use this algorithm in `Fake.DotNet.Cli` module? Specifically in the `findPossibleDotnetCliPaths` method https://github.com/fsprojects/FAKE/blob/c7273c70ff688d1e53d2b56809724ff55e3d9961/src/app/Fake.DotNet.Cli/DotNet.fs#L81-L97

cc: @matthid , @baronfel

File: src/app/Fake.Runtime/SdkAssemblyResolver.fs (hunk 1 above)
Comment (baronfel) [reply src/app/Fake.Runtime/SdkAssemblyResolver.fs:217]: agree here, we can't assume the presence of global.json. the safest path is to invoke the `dotnet` found from the new code you added and call `--version`. Proj-Info also does this [here](https://github.com/ionide/proj-info/blob/main/src/Ionide.ProjInfo/Library.fs#L101-L110)

**Comment by baronfel** [comment]
@yazeedobaid yes, but with one modification - the current algorithm supports checking the user-local install directories if as a last-resort, but the code from proj-info doesn't. I think that behavior should be kept for probing purposes, and TBH it should be added to proj-info as well. e.g. adding an ` |> Option.orElseWith tryFindFromDefaultUserDirs` to [this lazy code](https://github.com/ionide/proj-info/blob/main/src/Ionide.ProjInfo/Utils.fs#L83-L84)

**Comment by mclark1129** [comment]
Guessing this is related, although not entirely the same as the original issue mentioned on the PR: https://github.com/fsprojects/FAKE/issues/2641

**Comment by yazeedobaid** [comment]
@nojaf I have merged #2640 can you please update our branch from `release/next`

Thanks

**Comment by nojaf** [comment]
@yazeedobaid rebased.

**Comment by yazeedobaid** [comment]
Thanks @nojaf 
For using this algorithm in `Fake.DotNet.Cli` module as well, I have opened an issue for it #2644 we can track it in that issue.
I'll merge this now and release it

Thanks again

File: src/app/Fake.Runtime/SdkAssemblyResolver.fs (hunk 1 above)
Comment (mclark1129) [reply src/app/Fake.Runtime/SdkAssemblyResolver.fs:217]: Why doesn't `dotnet --version` work for all cases?

**Pull request merged**